### ISSUE TITLE
Negative peak frequency

### ISSUE THREAD

**Issue by benoitp-cmc** [NOAA-EMC/WW3#304]
On occasion one of our wave models has been giving negative or very large (100 s) values of peak period (and corresponding negative or very small peak frequency) at the odd grid point. It happens every few weeks. 

I've traced it back to w3iogomd:

- FP0 ends up too small because EL is negative, that is when ```EBD(ICEN,JSEA) < EBD(ILOW,JSEA)```. 
- The anomalous cases I've looked at had: IKP0=2 so ILOW=1, ICEN=2, IHGH=3. 
- The calculation of IKP0 does not allow IKP0=1 or IKP0=NK. 
- When I changed https://github.com/NOAA-EMC/WW3/blob/4dfafdbed48f42b3c7314693c55be758b5743c51/model/ftn/w3iogomd.ftn#L1947 to ```DO IK=NK-1, 1, -1``` I got reasonable peak period (29 s). 
- To be safe, one should also add ```.AND. IK .GT. 1``` in the ST1/ST3/ST4 conditionals to avoid checking for ```EBD(IK-1, JSEA)```.

**Is changing the DO loop this way a bad idea?** Is there a better way.

If it's a good idea, should the loop be modified more to allow IKP0 = NK?

Our first frequency is 0.035, maybe we should lower it a little.

#### Extra information on synoptic situation

The latest case was of a forecast initialized on 2021-01-27 at 12Z. The atmospheric model had a very strong storm at hour 66 (Hs at 20 m) around 40N 50W. This is unusually far east in the Atlantic for such a strong storm. In the mean time, NW Iceland was not getting much waves because they were so close to the ice edge off Greenland. As a result, at hour 107 and 110 each a coastal point off NW Iceland the peak period was from the very long swells from the storm.

Similar case on a forecast initialized on 2020-02-11 at 00Z at hour 116. This time the forecast big storm was around 50N 30W and the negative peak period at a point on the ice edge near Spitsbergen.

An earlier case (2018-11-25) the same happened at a coastal point near Málaga, Spain. Again it was with a big storm over the middle Atlantic sending large swells. I'm impressed the angle was right enough for this swell to make it through the Gibraltar Strait to Málaga.

I guess this happens more rarely in the other basins because strong enough storms are rarer and/or there's too much background swell.

**Comment by benoitp-cmc** [comment]
To answer myself (informed disagreements are more than welcome).

The lowest frequency should be allowed as the peak frequency.

I am confident changing this loop only affects FP0 and THP0 (peak frequency and peak direction).

I get the impression there's not much point allowing IKP0 = NK. If there is no wave energy then having IKP0=0 and FP0=UNDEF is fine. The only risk of not allowing it is if ```EBD(ICEN,JSEA) < EBD(IHGH,JSEA)``` when IKP0=NK-1, I don't know what it is but I guess there is a reason this cannot happen.

**Comment by benoitp-cmc** [comment]
Just changing the loop has an issue. When IKP0=1 we automatically have EL=0 and I believe the continuous peak period formula becomes an extrapolation (it does give me peak periods greater than my highest discrete peak period):
https://github.com/NOAA-EMC/WW3/blob/4dfafdbed48f42b3c7314693c55be758b5743c51/model/ftn/w3iogomd.ftn#L2010

To avoid an extrapolation we could disable the continuous formula when IKP0=1 so that we simply have FP0=SIG(1)/2*pi. 

Unless someone knows how to fix the continuous formula to not degenerate when IKP0=2 and EBD(1) > EBD(2).

**Comment by aliabdolali** [comment]
Hi @benoitp-cmc 
I am curious to know what was the mean period for those extreme cases? Can we correlate it to the peak frequency or ~5 sqrt(Hm0)  in such a condition with unrealistic values?
I don't think it is a good idea to have the peak frequency extrapolated outside [min-max] frequency range, also not =min or max frequency. I am just thinking out loud :)

**Comment by benoitp-cmc** [comment]
Thanks @aliabdolali 

I have output some point data. At:
https://hpfx.collab.science.gc.ca/~bpo001/WW3/issue_304/

Two points of interest
- IS1C (-24.000  65.750) for forecast hour 107: 20210131 230000
- IS2C (-20.500  65.750) for forecast hour 110: 20210201 020000

Here's the bulk parameters for IS1C at hour 107:

- HS=0.70 m
- 1/FP0=-26.5 s
- T02=3.6 s
- T0M1=10.7 s
- T01=5.4 s
- PHS(0)=0.30 m  PTP(0)=2.2 s
- PHS(1)=0.45 m  PTP(1)=11.2 s
- PHS(2)=0.26 m  PTP(2)=14.1 s

(As an aside, we are talking small waves so it may not be significant, but the square root of the sum of the square of the partition heights is 0.60 m, this is pretty far from HS=0.70m. A fourth and fifth partition around 0.25 m would not be enough to get the sum up there, but one swell at 0.36 m would do it. A neighbouring point has about that missing as well. Another neighbouring point with PTP(2)=27.4 s has PHS(2)=0.5m and HS matches well to the square root of the sum of the squares. That suggest the energy in the minimum frequency band might not be taken into account by the partition algorithm.)


I have trouble with the idea of not allowing the min/max discrete frequency as a valid peak. In my IS1C example, the minimum frequency has the most energy by a fair margin. In the corresponding point output (mparams), the peak period for that point is 29.93 s, this is near 1.05 times the maximum peak period in my model where I have 1.1 frequency steps. I am not too keen on extrapolating but constrained extrapolation like this doesn't seem excessive.

I guess I should look at the calculation for point output and see if it can be applied here. But not tonight.

**Comment by benoitp-cmc** [comment]
The peak period calculation in ww3_outp and ww3_ounp is very similar to that in w3iogomd. There are two differences, in the point output:

- The loop on IK to find the peak frequency goes from NK-1 to 1
- There is a check on HSMIN (set as 0.05m)

The problem if we allow IKP0=1 with no other change is that the peak frequency function is not continuous there. With IKP0=1 the function no longer depends on EL and EH and we always get the minimum peak frequency times 1+0.5*(1-1/XFR). In my case, if IKP0=1 I always get 29.93s, otherwise the highest possible peak frequency is 27.21s (with IKP0=2 and equal energy in the two smallest frequencies).

- If we allow IKP0=1 we need to change the interpolation function for this case to something that won't degenerate.
- If we don't allow IKP0=1, we need to add a conditional so that if EL > 0 then set EL=0.

**Comment by JessicaMeixner-NOAA** [comment]
Wait... the peak period calculation in ww3_outp and ww3_ounp should be using the same calculations as each other... and as that in w3iogo... we should fix that.

**Comment by benoitp-cmc** [comment]
ww3_outp and ww3_ounp are the same as far as I can tell. Different from w3iogomd and that is a problem. As it is, I think both need improvement.

The existence of the min and max function to get ILOW and IHGH tells me the original intention was to allow the IK loop to go from NK to 1 rather than NK-1 to 2:
https://github.com/NOAA-EMC/WW3/blob/4dfafdbed48f42b3c7314693c55be758b5743c51/model/ftn/w3iogomd.ftn#L2004

However, those min/max are not the right solution. They make sure we don't pick EBD from a random address in memory but they ultimately can lead to the problem I observed with peak frequencies that are too small.

To me the solution is to allow IK from NK to 1 and remove the calculation of IHGH, ICEN and ILOW and instead calculate EL and EH with extra conditionals (compiles but not tested yet, I'll try to test today):
```
        IF ( IKP0(JSEA) .EQ. 1 ) THEN
          EL = - EBD(IKP0(JSEA), JSEA)
        ELSE
          EL = EBD(IKP0(JSEA)-1, JSEA) - EBD(IKP0(JSEA), JSEA)
        END IF
        IF ( IKP0(JSEA) .EQ. NK ) THEN
          EH = - EBD(IKP0(JSEA), JSEA)
        ELSE
          EH = EBD(IKP0(JSEA)+1, JSEA) - EBD(IKP0(JSEA), JSEA)
        END IF
```
This should allow peak frequency to equal the minimum and maximum discrete frequency, be continuous in that whole range, and not allow any extrapolation.

**Comment by benoitp-cmc** [comment]
I am satisfied with the test result.

@JessicaMeixner-NOAA @aliabdolali Can we safely assume the second highest frequency always has at least as much energy as the highest frequency? 

If yes, we can have the IK loop from NK-1 to 1 rather than adding NK, and we won't need the conditional IF (IKP0 .EQ. NK).

**Comment by benoitp-cmc** [comment]
I think I have a good fix here: 14d72c3c13db81488b638d029b692107980deed6

I had missed it but outp/ounp allow peak frequency to be IK=NK through the initial value. The suggested fix makes my question in the previous comment moot (the answers appears to be we cannot assume).

The calculation is the same between gridded and point output except I would leave the use of HSMIN in ww3_outp/ww3_ounp. In w3iogomd I have left no restriction on energy.

I imagine it will break some regression tests...

**Comment by JessicaMeixner-NOAA** [comment]
@benoitp-cmc I agree that it makes more sense to not make assumptions.   So with the commit you made would that make all of the peak frequency calculations consistent?   I agree this will change answers of regression tests most likely.

**Comment by benoitp-cmc** [comment]
@JessicaMeixner-NOAA the calculation is consistent between point and field output except for the check on HSMIN in point output. 

The HSMIN check applies to many fields in point output, here I'd keep it on FP/TP to be consistent. I'm not clear on its usefulness (I've disabled it in the past to debug something or other), but I have created another issue (#311) to ask opinions on if we should keep or disable it (setting HSMIN to 0 would do it, or more flexible would be to read from a namelist).
